### PR TITLE
Add CUDA kernel for prior_box_op.

### PR DESCRIPTION
**Pull request by qingqing01**
Related to https://github.com/PaddlePaddle/Paddle/issues/9472

### CODE REVIEW

File: paddle/fluid/operators/prior_box_op.cu [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+template <typename T>
+__global__ void GenPriorBox(T* out, const T* aspect_ratios, const int height,
+                            const int width, const int im_height,
+                            const int im_width, const int as_num,
+                            const T offset, const T step_width,
+                            const T step_height, const T* min_sizes,
+                            const T* max_sizes, const int min_num,
+                            bool is_clip) {
+  int num_priors = max_sizes ? as_num * min_num + min_num : as_num * min_num;
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/prior_box_op.cu]: 你在[host上](https://github.com/PaddlePaddle/Paddle/pull/9553/files#diff-6e391bc8619b4ad8145423606a3b08faR120)不是算过`num_priors`了么?

而且两个地方的计算方式貌似不一样，确认下？

File: paddle/fluid/operators/prior_box_op.cu [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  int box_num = height * width * num_priors;
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < box_num;
+       i += blockDim.x * gridDim.x) {
+    int h = i / (num_priors * width);
+    int w = (i / num_priors) % width;
+    int p = i % num_priors;
+    int m = max_sizes ? p / (as_num + 1) : p / as_num;
+    T cx = (w + offset) * step_width;
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/prior_box_op.cu]: 第一个box就需要乘上step_width么？

Comment (qingqing01) [reply paddle/fluid/operators/prior_box_op.cu]: Yeah, each box needs to mul step_width https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/PriorBox.cpp#L105

File: paddle/fluid/operators/prior_box_op.cu (hunk 1 above)
Comment (qingqing01) [reply paddle/fluid/operators/prior_box_op.cu]: 想少传递一些参数，就在kernel里也计算了下。
max_sizes 和 min_sizes的个数是相等的。 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/prior_box_op.cc#L52 这里有check。 所以这里就直接 + min_num。

File: paddle/fluid/operators/prior_box_op.cu [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/operators/prior_box_op.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename T>
+__device__ inline T clip(T in) {
```
Comment (chengduoZH) [inline paddle/fluid/operators/prior_box_op.cu]: `clip` -> `Clip`

**Review by chengduoZH — COMMENTED** [review]
Excellent!
My biggest concern is its performance, and I feel that there are some points to be improved.

**Review by chengduoZH — APPROVED** [review]

**Pull request merged**